src/db/cow/mod.rs

```diff
@@ -0,0 +1,231 @@
+            return Ok(account);
+        }
+        self.inner.basic_ref(address)
+    }
+
+    fn code_by_hash(&mut self, code_hash: B256) -> Result<Bytecode, Self::Error> {
+        if let Some(code) = self.cache.contracts.get(&code_hash) {
+            return Ok(code.clone());
+        }
+        self.inner.code_by_hash_ref(code_hash)
+    }
+
+    fn storage(&mut self, address: Address, index: U256) -> Result<U256, Self::Error> {
+        if let Some(storage) =
+            self.cache.accounts.get(&address).map(|a| a.storage.get(&index).cloned())
+        {
+            return Ok(storage.unwrap_or_default());
+        }
+        self.inner.storage_ref(address, index)
+    }
+
+    fn block_hash(&mut self, number: u64) -> Result<B256, Self::Error> {
+        if let Some(hash) = self.cache.block_hashes.get(&U256::from(number)) {
+            return Ok(*hash);
+        }
+        self.inner.block_hash_ref(number)
+    }
+}
+
+impl<Db: DatabaseRef> DatabaseRef for CacheOnWrite<Db> {
+    type Error = Db::Error;
+
+    fn basic_ref(&self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        if let Some(account) = self.cache.accounts.get(&address).map(DbAccount::info) {
+            return Ok(account);
+        }
+        self.inner.basic_ref(address)
+    }
+
+    fn code_by_hash_ref(&self, code_hash: B256) -> Result<Bytecode, Self::Error> {
+        if let Some(code) = self.cache.contracts.get(&code_hash) {
+            return Ok(code.clone());
+        }
+        self.inner.code_by_hash_ref(code_hash)
+    }
+
+    fn storage_ref(&self, address: Address, index: U256) -> Result<U256, Self::Error> {
+        if let Some(storage) =
+            self.cache.accounts.get(&address).map(|a| a.storage.get(&index).cloned())
+        {
+            return Ok(storage.unwrap_or_default());
+        }
+        self.inner.storage_ref(address, index)
+    }
+
+    fn block_hash_ref(&self, number: u64) -> Result<B256, Self::Error> {
+        if let Some(hash) = self.cache.block_hashes.get(&U256::from(number)) {
+            return Ok(*hash);
+        }
+        self.inner.block_hash_ref(number)
+    }
+}
+
+impl<Db> DatabaseCommit for CacheOnWrite<Db> {
+    fn commit(&mut self, changes: HashMap<Address, Account>) {
+        for (address, mut account) in changes {
+            if !account.is_touched() {
+                continue;
+            }
+
+            if account.is_selfdestructed() {
+                let db_account = self.cache.accounts.entry(address).or_default();
+                db_account.storage.clear();
+                db_account.account_state = AccountState::NotExisting;
+                db_account.info = AccountInfo::default();
+                continue;
+            }
+
+            let is_newly_created = account.is_created();
+            self.cache.insert_contract(&mut account.info);
+
+            let db_account = self.cache.accounts.entry(address).or_default();
+            db_account.info = account.info;
```

Comment:
I assume this is safe even in the `SelfDestructed` case?

Reply:
this code is reproduced from revm directly. which reminds me i need to add a note and a copyright notice

Reply:
done 🫡